### PR TITLE
RavenDB-22972 Dequeued EnvirnmentStateRecord elements should be explictly nulled because they might hold references of objects which have finalizers (ClientState)

### PR DESCRIPTION
**Pull request by arekpalinski**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22972/Rising-managed-memory-usage-and-GC-impact-on-7.0-cluster

### Additional description

See details in the following comment: https://issues.hibernatingrhinos.com/issue/RavenDB-22972/Rising-managed-memory-usage-and-GC-impact-on-7.0-cluster#focus=Comments-67-1056937.0-0

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [x] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Voron/StorageEnvironment.cs [hunk 1]
```diff
@@ -1687,16 +1687,24 @@ internal ApplyLogsToDataFileState TryGetLatestEnvironmentStateToFlush(long uptoT
             while (true)
             {
                 if (_transactionsToFlush.TryPeek(out var maybe) == false ||
-                    maybe.TransactionId >= uptoTxIdExclusive)
+                    maybe.EnvStateRecord.TransactionId >= uptoTxIdExclusive)
```
Comment (ayende) [inline src/Voron/StorageEnvironment.cs:1695]: Can this be null?

File: src/Voron/StorageEnvironment.cs [hunk 2]
```diff
@@ -1340,11 +1340,11 @@ public InMemoryStorageState GetInMemoryStorageState(LowLevelTransaction tx)
                 },
                 TransactionsToFlush = _transactionsToFlush.ToList().Select(x => new InMemoryStorageState.EnvironmentStateRecordDetails
                 {
-                    TransactionId = x.TransactionId,
-                    ScratchPagesTable = GetScratchTableSummary(x.ScratchPagesTable),
-                    NextPageNumber = x.NextPageNumber,
-                    WrittenToJournalNumber = x.WrittenToJournalNumber,
-                    ClientStateType = x.ClientState?.GetType().ToString()
+                    TransactionId = x.EnvStateRecord.TransactionId,
```
Comment (ayende) [inline src/Voron/StorageEnvironment.cs]: Let's access this just once?

**Review by ayende — COMMENTED** [review]
Amazing catch!

File: src/Voron/StorageEnvironment.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Voron/StorageEnvironment.cs:1695]: No, we set it to `null` _after_ dequeuing. So as long as it's in the queue it must not be `null`.

File: src/Voron/StorageEnvironment.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Voron/StorageEnvironment.cs]: Yeah, I will fix it.

**Review by ayende — COMMENTED** [review]
I wonder if we instead of adding the additional indirection, we can skip the `TryPeek` and keep a field for the `_lastDequeue`, that will serve as the same thing.

**Pull request merged**